engine/kv_engine.cpp

```diff
@@ -1496,15 +1512,169 @@ Status KVEngine::SGet(const StringView collection, const StringView user_key,
+                                           ExpiredTimeType expired_time) {
+  auto value = skiplist->header()->record->Value();
+  auto key = skiplist->header()->record->Key();
+  auto request_size = value.size() + key.size() + sizeof(DLRecord);
+  SpaceEntry sized_space_entry = pmem_allocator_->Allocate(request_size);
+  if (sized_space_entry.size == 0) {
+    return Status::PmemOverflow;
+  }
+  void* new_head_record =
+      pmem_allocator_->offset2addr(sized_space_entry.offset);
+
+  while (1) {
+    SkiplistNode* dram_node = nullptr;
+    HashEntry* entry_ptr = nullptr;
+    HashEntry hash_entry;
+    DataEntry data_entry;
+    DLRecord* existing_record = nullptr;
+    auto hint = hash_table_->GetHint(key);
+    std::unique_lock<SpinMutex> ul(*hint.spin);
+    version_controller_.HoldLocalSnapshot();
+    defer(version_controller_.ReleaseLocalSnapshot());
+    TimeStampType new_ts =
+        version_controller_.GetLocalSnapshot().GetTimestamp();
+    Status s = hash_table_->SearchForRead(hint, key, SortedHeaderRecord,
+                                          &entry_ptr, &hash_entry, &data_entry);
+    if (s != Status::Ok) return s;
+    dram_node = entry_ptr->GetIndex().skiplist->header();
+    existing_record = dram_node->record;
+    if (!skiplist->Update(key, value, existing_record, hint.spin, new_ts,
+                          RecordType::SortedHeaderRecord, dram_node,
+                          sized_space_entry, expired_time)) {
+      continue;
+    }
+
+    hash_table_->Insert(hint, entry_ptr, SortedHeaderRecord, skiplist,
+                        HashIndexType::Skiplist);
+    //(TODO) free head record
+    break;
+  }
+  return Status::Ok;
+}
+
+Status KVEngine::InplaceUpdatedExpiredTime(const StringView& str,
+                                           ExpiredTimeType expired_time,
+                                           RecordType record_type) {
+  HashTable::KeyHashHint hint = hash_table_->GetHint(str);
+  std::unique_lock<SpinMutex> ul(*hint.spin);
+  HashEntry hash_entry;
+  HashEntry* entry_ptr = nullptr;
+  Status s = hash_table_->SearchForRead(hint, str, record_type, &entry_ptr,
+                                        &hash_entry, nullptr);
+  if (s != Status::Ok) return s;
+  switch (entry_ptr->GetIndexType()) {
+    case HashIndexType::Queue:
+      entry_ptr->GetIndex().queue_ptr->InplaceUpdatedExpiredTime(expired_time);
+      break;
+    case HashIndexType::UnorderedCollection:
+      entry_ptr->GetIndex().p_unordered_collection->InplaceUpdatedExpiredTime(
+          expired_time);
+      break;
+    default:
+      return Status::NotSupported;
+  }
+  return Status::Ok;
+}
+
+Status KVEngine::SetExpiredTime(const StringView str,
+                                ExpiredTimeType expired_time) {
+  if (!CheckOverFlow(expired_time)) {
+    return Status::InvalidArgument;
+  }
+
+  HashTable::KeyHashHint hint = hash_table_->GetHint(str);
+  HashEntry hash_entry;
```

Comment:
You should lock, SearchForRead() and update that time here, instead of search twice.

Reply:
It's a write operation, you should lock here.